src/VisualStudio/Core/Impl/ProjectSystem/CPS/TempPECompiler.cs

```diff
@@ -37,22 +40,25 @@ public async Task<bool> CompileAsync(IWorkspaceProjectContext context, string ou
                 throw new ArgumentException(nameof(outputFileName), "Must specify an output file name.");
             }
 
-            var project = _workspace.CurrentSolution.GetProject(context.Id);
+            var project = _workspace.CurrentSolution.GetRequiredProject(context.Id);
 
-            // Remove all files except the ones we care about
-            var documents = project.Documents;
-            foreach (var document in documents)
+            // Remove all files except the ones we care about. We do this in a batch so we aren't removing syntax trees from the compilation
+            // one at a time.
+            var documentIdsToRemove = ImmutableArray.CreateBuilder<DocumentId>(initialCapacity: project.DocumentIds.Count);
+            foreach (var document in project.Documents)
             {
-                if (!filesToInclude.Contains(document.FilePath))
+                if (document.FilePath != null && !filesToInclude.Contains(document.FilePath))
                 {
-                    project = project.RemoveDocument(document.Id);
+                    documentIdsToRemove.Add(document.Id);
                 }
 
                 cancellationToken.ThrowIfCancellationRequested();
             }
 
+            project = project.RemoveDocuments(documentIdsToRemove.ToImmutable());
```

Comment:
@davidwengier: is the common case that we only end up with one or two files left? @agocke pointed out that we could just grab the compilation, call RemoveAllSyntaxTrees(), and then add the ones you want back in. It would have meant I could have avoided the rest of the work, but we did want that done anyways.

Reply:
Yes, the common case is a very small set of files, maybe as high as 3 if you want to be conservative. I had a look at this and thought a RemoveAllDocuments would make sense since there was already a RemoveAllSyntaxTrees, but that was just a guess based on convenience 😃

Reply:
Alright in that case then I'm going to just make that change directly in a separate PR and I'll deal with the fancier API stuff later.

Reply:
https://github.com/dotnet/roslyn/pull/41292 is now the direct implementation of that approach. This PR will still go through for the other perf benefits, but it's not related to TempPE anymore.